### PR TITLE
Ehn/rpos

### PR DESCRIPTION
**Pull request by mts299**
# Scope 

Converts range gate locations to geographic coordinates 

**issue:** #141 

## Approval

**Number of approvals:** 2 - 1 for testing 1 for review but more might be nicer. 

## Test

```python
import pydarn

lat, lon = pydarn.geographic_cell_positions(66, 3, 60, 45, 180, height=300)
print(lat)
print(lon)
```
Output: 
```bash
63.79079100452335
-134.3697504532961
```

To check you can look at the fov files, or try to use RST/IDL code or https://superdarn.ca/radar-coverage which is a handy tool for this feature.

### CODE REVIEW

**Comment by mts299** [comment]
This is daft for the slant range code and I need to add this FOV code for fan plots and tests speed oof 
- [ ] remove slant range function in the code and use pyDARN's 
- [ ] update const file 
- [ ] add to FOV code 
- [ ] performance test

**Comment by mts299** [comment]
Update for plot_fov to work with this code. Try it out!

``` python
import pydarn
import matplotlib.pyplot as plt 
from datetime import datetime

fitacf_file = 'data/20210205.1200.00.cly.a.fitacf'
fitacf_data = pydarn.SuperDARNRead().read_dmap(fitacf_file)
pydarn.Fan.plot_fov(66, datetime(2021, 2, 5, 12, 5)) 
pydarn.Fan.plot_fov(66, datetime(2021, 2, 5, 12, 5), fov_files=True)
plt.show()
```

Still working on `plot_fan`

**Comment by mts299** [comment]
plot_fan now works!! 

Figure_1

Test it out, I do need to clean up some of the code.

**Review by Devin-Huyghebaert — COMMENTED** [review]
> Update for plot_fov to work with this code. Try it out!
> 
> ```python
> import pydarn
> import matplotlib.pyplot as plt 
> from datetime import datetime
> 
> fitacf_file = 'data/20210205.1200.00.cly.a.fitacf'
> fitacf_data = pydarn.SuperDARNRead().read_dmap(fitacf_file)
> pydarn.Fan.plot_fov(66, datetime(2021, 2, 5, 12, 5)) 
> pydarn.Fan.plot_fov(66, datetime(2021, 2, 5, 12, 5), fov_files=True)
> plt.show()
> ```
> 
> Still working on `plot_fan`

Tested the plot_fov with the example above (on the right of the figure below).  There appears to be slight discrepancies between the files and the new rpos code.  Looks like a shift of the range gates potentially? The transition from E- to F-region also appears to be slightly different.

plot_fov_example

Also checked the plot_fan function.  Looks good to me, but I haven't dove into the numbers yet.  As a rough check, it appears similar to the plots generated from the VT website.

plot_fan_example
vt_map_example

Some other small questions/suggestions for the code are below.  Overall looks good!

File: pydarn/utils/radar_pos.py [hunk 1]
```diff
@@ -78,3 +132,434 @@ def radar_fov(stid: int, coords: str = 'aacgm', date: datetime = None):
...
+
+    # TODO: fix after slant range change
+    if center is True:
+        # beam edge in [rad]
+        beam_edge = -beam_sep * 0.5
+        # range_edge in [km]
+    else:
+        beam_edge = 0
+
+    # psi in [rad]
```
[the remark this answers is not in the document]
Comment (mts299) [reply pydarn/utils/radar_pos.py:207]: Good question, @ecbland ?

File: pydarn/utils/radar_pos.py [hunk 2]
```diff
@@ -78,3 +132,434 @@ def radar_fov(stid: int, coords: str = 'aacgm', date: datetime = None):
...
+    psi = beam_sep * (beam - offset) + beam_edge
+    # Calculate the slant range [km]
+    slant_range = gate2slant(frang, rsep, rxrise, gate=range_gate)
+
```
[the remark this answers is not in the document]
Comment (mts299) [reply pydarn/utils/radar_pos.py:211]: The elevation is used by default stated in the above parameters. Unless some one goes None of both which then python will fail

File: pydarn/utils/radar_pos.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ecbland) [reply pydarn/utils/radar_pos.py:207]: `psi` is the angle from the boresight. Here are the equivalent lines in RST: https://github.com/SuperDARN/rst/blob/b667628c85580b9bbcff6b167345a33a5d0c1c56/codebase/superdarn/src.lib/tk/rpos.1.7/src/cnvtcoord.c#L360-L361

**Comment by mts299** [comment]
Profiled it with C-profile and got it to plot with rpos code with 1.24 seconds vs. reading a file 1.14 seconds. 

We can discuss this in the next pyDARN meeting but does this 0.10 seconds allow us to deprecate the fov files assuming this code is accurate?

**Review by carleyjmartin — COMMENTED** [review]
I'm getting an error for FOV, fan and grid plots:
```
Traceback (most recent call last):
  File "fov_plot.py", line 6, in <module>
    fanplot=pydarn.Fan.plot_fov(66, datetime(2021, 2, 5, 12, 5), radar_location=True, radar_label=True,  lowlat=50, boundary=True, line_color='red')
  File "/Users/carley/Documents/testing_code/venv-rpos/lib/python3.7/site-packages/pydarn/plotting/fan.py", line 333, in plot_fov
    radar_fov(stid, rsep, frang, coords=coords, ranges=ranges,
NameError: name 'rsep' is not defined
```
I'm guessing it's probably something come up when merging all the new changes in, probably a kwargs issue? Might need to have a default set somewhere.

Code used:
```python
import matplotlib.pyplot as plt
import pydarn
import datetime as dt

#Grid
grid_single = "/Users/carley/Documents/data/grids/20011220.0800.00.han.grid"
grid_data = pydarn.SuperDARNRead(grid_single).read_grid()
pydarn.Grid.plot_grid(grid_data, parameter = 'pwr', radar_label=True, 
                      radar_location=True,fov_color='grey', lowlat=50)
plt.show()

#Fan
file_north = "/Users/carley/Documents/data/20131031.2201.00.sas.fitacf"
SDarn_read = pydarn.SuperDARNRead(file_north)
time = dt.datetime(2013,10,31,22,4)
fitacf_north = SDarn_read.read_fitacf()
fanplot = pydarn.Fan.plot_fan(fitacf_north, scan_index=time,
                    colorbar_label='Velocity [m/s]',
                    line_color='red', radar_label=True)
plt.show()

# FOV
fanplot=pydarn.Fan.plot_fov(66, dt.datetime(2021, 2, 5, 12, 5), radar_location=True, radar_label=True,  lowlat=50, boundary=True, line_color='red')
```

**Comment by mts299** [comment]
Yea was going to fix it up with all the conflicts

**Comment by carleyjmartin** [comment]
Works fine for plotting, except when using lowlat kw for FOV, fan and grids:
```
Traceback (most recent call last):
  File "release_test.py", line 20, in <module>
    radar_location=True,fov_color='grey',lowlat=50)
  File "/Users/carley/Documents/testing_code/venv-rpos/lib/python3.7/site-packages/pydarn/plotting/grid.py", line 164, in plot_grid
    ax=ax, **kwargs)
  File "/Users/carley/Documents/testing_code/venv-rpos/lib/python3.7/site-packages/pydarn/plotting/fan.py", line 328, in plot_fov
    radar_fov(stid, ranges=ranges, date=date, **kwargs)
TypeError: radar_fov() got an unexpected keyword argument 'lowlat'
```

**Review by carleyjmartin — APPROVED** [review]
Everything works well for me now! I had a quick read through radar_pos.py but as I've seen it before and was involved in the writing it's probably best if a fresh pair of eyes looks that over.

**Pull request merged**